### ISSUE TITLE
Sail commands don't exit correctly on quit

### ISSUE THREAD

**Issue by dshafik** [soloterm/solo#42]
When running sail commands e.g. `sail up` inside Solo (Solo is running on the host, not inside sail), and quitting Solo, the `sail` command don't terminate. 

This means that a command of `sail npm run dev` leaves `vite` running inside the container and when starting Solo again it won't start as the port is already in use.

I suspect that each command needs to be sent an explicit `SIGTERM` or `SIGKILL` when `q` is pressed inside the Solo UI.

**Comment by dshafik** [comment]
I have figured out why this is happening… `docker compose` doesn't kill the process in the container on `SIGTERM`, only the `docker` command itself it seems. I think this is because when you do a Ctrl+C in the terminal, that is being forwarded to the _underlying_ `npm` (or whatever) command which then exits and the docker process dies automatically.

You get this exact same behavior if you run `docker compose exec` with the `-T` flag (e.g. `docker compose exec -T <container> <command>`):

```
  -T, --no-TTY docker compose exec   Disable pseudo-TTY allocation. By default docker compose exec allocates
                                     a TTY.
```

To be clear, this isn't _really_ an issue with Solo, but it should be documented that commands that expect a TTY (like docker/sail) cannot be used with Solo.

**Comment by aarondfrancis** [comment]
Hmm... I don't use Sail so I don't fully understand what's happening here. So when Solo tries to kill the process it just kills the docker compose process on the host machine and not the process inside of docker? Is there a way to kill that underlying process? It'd be nice if it just worked with Solo

**Comment by dshafik** [comment]
> So when Solo tries to kill the process it just kills the docker compose process on the host machine and not the process inside of docker?

Exactly. This isn't unique to sail but illustrative of a problem with a common Laravel tool lots of folks will be familiar with .

> Is there a way to kill that underlying process? It'd be nice if it just worked with Solo

There are no easy solutions here, you either need to implement a virtual TTY in PHP, or spawn the process under another tool that does so.

**Comment by aarondfrancis** [comment]
All processes [are now spawned with a PTY](https://github.com/soloterm/solo/blob/main/src/Commands/Concerns/ManagesProcess.php#L56). Does this get us any closer?

**Comment by dshafik** [comment]
@aarondfrancis This doesn't solve the issue in this ticket. This arguable makes things **worse**, any command that can take input will now stop solo from receiving the keyboard input (e.g. `sail up` and then you can't change tabs with the arrow keys).

**Comment by aarondfrancis** [comment]
Haha dang. Well I'd you have any ideas let me know!

**Comment by dshafik** [comment]
@aarondfrancis I would roll back this change ASAP

**Comment by aarondfrancis** [comment]
Sorry, that's not viable at this point. If you can get me a reproduction I can look into it. I don't use sail though so you'll need to be explicit

**Comment by dshafik** [comment]
@aarondfrancis any command that expects input will work, so `sudo echo "Hello World"` should do it (so long as it asks for your password)

**Comment by aarondfrancis** [comment]
Cool, thank you!

**Comment by aarondfrancis** [comment]
Ok I've got the `sudo echo "Hello World"` example working in #50, but I haven't had a chance to test it very thoroughly. I put `screen` in the middle and that seems to be working quite nicely!

**Comment by aarondfrancis** [comment]
@dshafik can you composer install that branch reference and see if it works for Sail?

**Comment by dshafik** [comment]
@aarondfrancis this is way closer! With `sail up` (basically `docker compose up`) it doesn't allow input but it doesn't block moving between tabs (this is fine!). With `sail npm run dev` it still doesn't kill correct so it won't restart, **but** if I use `sail npm run dev && sail exec beacon killall node` as the command I can hit restart twice and it works, which is a big step forward.

It also solved an issue running `sail pest` (it would break the frame and then redraw after it was done, I suspect similar to ngrok that you mentioned had issues before).

At this point it's totally usable for me.

**Comment by aarondfrancis** [comment]
We're getting somewhere! If you want to allow input, you can try 

`Command::from('sail up')->interactive()` and then hit <kbd>i</kbd> to go interactive.

**Comment by aarondfrancis** [comment]
@dshafik I just a new release. Will you give it a go an let me know what you think? And if all is well can we close this one?
